### PR TITLE
fix(transformer): test cases migration

### PR DESCRIPTION
**Pull request by mihir-4116**
## Description of the change

- This pull request involves migrating existing test cases (proxy, user deletion, and router) to TypeScript. 

- Additionally, 3  new unit test commands introduced

1. `npm run test:js` : to run all the javascript file test cases
2. `npm run test:ts` : to run all the typescript file test cases
3. `npm run test:ut:integration` : to run user transformation integration test cases

- for all the test cases `:silent` and `:ci` flavours are added

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/controllers/destination.ts [hunk 1]
```diff
@@ -127,11 +127,11 @@ export default class DestinationController {
...
-    stats.histogram('dest_transform_output_events', ctx.body?.output?.length, {
-      destination,
-      version: 'v0',
-      ...metaTags,
-    });
+    // stats.histogram('dest_transform_output_events', ctx.body?.output?.length, {
+    //   destination,
+    //   version: 'v0',
+    //   ...metaTags,
+    // });
```
Comment (saikumarrs) [inline src/controllers/destination.ts:134]: Let's remove commented code, please.

File: src/controllers/util/index.ts [hunk 2]
```diff
@@ -13,7 +13,8 @@ export default class ControllerUtility {
   }
 
   public static deliveryPostProcess(ctx: Context, status = 200) {
-    ctx.set('apiVersion', API_VERSION);
+    // To Do(fix error) : ctx.set is not a function
```
Comment (saikumarrs) [inline src/controllers/util/index.ts:16]: Use `TODO`. This is a standard convention.

File: jest.config.integration.js [hunk 3]
```diff
@@ -136,7 +136,7 @@ module.exports = {
   // testLocationInResults: false,
 
   // The glob patterns Jest uses to detect test files
-  testMatch: ['**/__tests__/**/*.(integration.test).[jt]s?(x)'],
+  testMatch: ['**/__tests__/**/*.(test).[t]s?(x)', '**/deleteUsers/**/*.(test).[t]s?(x)'],
```
Comment (saikumarrs) [inline jest.config.integration.js:139]: How is this different from the `jest.config.js` config?
Why is `deleteUsers` test part of integration tests?

File: .github/workflows/build-push-docker-image.yml [hunk 4]
```diff
@@ -62,6 +62,7 @@ jobs:
         run: |
           docker run ${{ inputs.build_tag }} npm run test:ci
           docker run ${{ inputs.build_tag }} npm run test:integration
+          docker run ${{ inputs.build_tag }} npm run test:utils
```
Comment (saikumarrs) [inline .github/workflows/build-push-docker-image.yml:65]: Both these scripts should have a `:ci` flavor.

File: jest.config.integration.js (hunk 3 above)
Comment (saikumarrs) [reply jest.config.integration.js:139]: CC @utsabc

File: jest.config.utils.js [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    testEnvironment: 'node',
+
+    // Options that will be passed to the testEnvironment
+    // testEnvironmentOptions: {},
+
+    // Adds a location field to test results
+    // testLocationInResults: false,
+
+    // The glob patterns Jest uses to detect test files
+    testMatch: ['**/__tests__/**/*.(test).[t]s?(x)', '**/deleteUsers/**/*.(test).[t]s?(x)'],
```
Comment (saikumarrs) [inline jest.config.utils.js:139]: We're running delete users here also?

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: src/controllers/destination.ts (hunk 1 above)
Comment (mihir-4116) [reply src/controllers/destination.ts]: kept original code as @utsabc will raise pr soon to fix it

File: jest.config.integration.js (hunk 3 above)
Comment (mihir-4116) [reply jest.config.integration.js]: removed typescript test cases from this config

File: src/controllers/destination.ts (hunk 1 above)
Comment (saikumarrs) [reply src/controllers/destination.ts]: That'll be available in the version history.

File: jest.config.utils.js [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+    // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
+    testPathIgnorePatterns: ['/node_modules/', 'user_transformation.integration.test.js'],
```
Comment (saikumarrs) [inline jest.config.utils.js:142]: We can run all the TS test scripts with this (including integration).
```suggestion
    testMatch: ['test/**/*.test.ts?(x)'],
    
    // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
    testPathIgnorePatterns: ['/node_modules/'],
```

Comment (mihir-4116) [reply jest.config.utils.js]: updated

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -74,5 +74,5 @@ jobs:
       - name: Port Forwarding to port on gateway-external
         run: kubectl port-forward service/gateway-external 8080:8080 -n openfaas &
 
-      - name: Jest Integration test
-        run: npm run test:integration
+      - name: Transformation test
```
Comment (saikumarrs) [inline .github/workflows/test.yml:77]:
```suggestion
      - name: User Transformation Tests
```

File: jest.config.ut.js [hunk 8]
```diff
@@ -136,10 +147,10 @@ module.exports = {
   // testLocationInResults: false,
 
   // The glob patterns Jest uses to detect test files
-  testMatch: ['**/__tests__/**/*.(integration.test).[jt]s?(x)'],
+  testMatch: ['**/__tests__/**/user_transformation.test.js'],
```
Comment (saikumarrs) [inline jest.config.ut.js:150]: A whole config file just for a single test seems redundant. Let's either move this file to a directory dedicated to UT tests or use `npx jest <file_name>` script to run these tests.

File: .github/workflows/test.yml (hunk 7 above)
Comment (mihir-4116) [reply .github/workflows/test.yml]: updated

File: jest.config.ut.js (hunk 8 above)
Comment (mihir-4116) [reply jest.config.ut.js:150]: updated as per suggestion

File: package.json [hunk 9]
```diff
@@ -34,12 +34,19 @@
     "verify": "eslint . || exit 1; npm run test || exit 1",
     "test:testRouter": "jest testRouter --detectOpenHandles --coverage --notify --watchAll=false",
     "test": "jest --detectOpenHandles --coverage --notify --watchAll=false",
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply package.json]: yes, removed `test` command

File: package.json [hunk 10]
```diff
@@ -29,17 +29,19 @@
...
+    "test:js": "jest -c jest.default.config.js --detectOpenHandles",
+    "test:js:silent": "npm run test:js -- --silent",
+    "test:js:ci": "npm run test:js -- --coverage --expand --maxWorkers=50%",
+    "test:ts": "jest -c jest.config.typescript.js --detectOpenHandles",
+    "test:ts:silent": "npm run test:ts -- --silent",
+    "test:ts:ci": "npm run test:ts -- --expand --maxWorkers=50%",
+    "test:ut:integration": "jest \"user_transformation.integration.test.js\"  --detectOpenHandles --notify",
+    "test:ut:integration:silent": "npm run test:ut:integration -- --silent",
+    "test:ut:integration:ci": "npm run test:ut:integration -- --expand --maxWorkers=50%",
+    "pre-commit": "npm run test:ts:silent && npm run test:js:silent && npm run test:ut:silent && npx lint-staged",
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply package.json]: updated pre-commit command

File: package.json [hunk 11]
```diff
@@ -41,7 +41,7 @@
     "test:ut:integration": "jest \"user_transformation.integration.test.js\"  --detectOpenHandles --notify",
     "test:ut:integration:silent": "npm run test:ut:integration -- --silent",
     "test:ut:integration:ci": "npm run test:ut:integration -- --expand --maxWorkers=50%",
-    "pre-commit": "npm run test:ts:silent && npm run test:js:silent && npm run test:ut:silent && npx lint-staged",
+    "pre-commit": "npm run test:ts:silent && npm run test:js:silent && npm run test:ut:integration:silent && npx lint-staged",
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply package.json]: right, it will fail. removed it from pre-commit command

**Review by utsabc — APPROVED** [review]

**Review by saikumarrs — APPROVED** [review]

**Pull request merged**